neo4j/_async/bookmark_manager.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) "Neo4j"
+# Neo4j Sweden AB [https://neo4j.com]
+#
+# This file is part of Neo4j.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import annotations
+
+import typing as t
+from collections import defaultdict
+
+from .._async_compat.concurrency import AsyncCooperativeLock
+from .._async_compat.util import AsyncUtil
+from ..api import (
+    AsyncBookmarkManager,
+    Bookmarks,
+)
+
+
+T_BmSupplier = t.Callable[[t.Optional[str]],
+                          t.Union[Bookmarks, t.Awaitable[Bookmarks]]]
+T_BmConsumer = t.Callable[[str, Bookmarks], t.Union[None, t.Awaitable[None]]]
+
+
+def _bookmarks_to_set(
+    bookmarks: t.Union[Bookmarks, t.Iterable[str]]
+) -> t.Set[str]:
+    if isinstance(bookmarks, Bookmarks):
+        return set(bookmarks.raw_values)
+    return set(map(str, bookmarks))
+
+
+class AsyncNeo4jBookmarkManager(AsyncBookmarkManager):
+    def __init__(
+        self,
+        initial_bookmarks: t.Mapping[str, t.Union[Bookmarks,
+                                                  t.Iterable[str]]] = None,
+        bookmarks_supplier: T_BmSupplier = None,
+        bookmarks_consumer: T_BmConsumer = None
+    ) -> None:
+        super().__init__()
+        self._bookmarks_supplier = bookmarks_supplier
+        self._bookmarks_consumer = bookmarks_consumer
+        if initial_bookmarks is None:
+            initial_bookmarks = {}
+        self._bookmarks = defaultdict(
+            set, ((k, _bookmarks_to_set(v))
+                  for k, v in initial_bookmarks.items())
+        )
+        self._lock = AsyncCooperativeLock()
+
+    async def update_bookmarks(
+        self, database: str, previous_bookmarks: t.Iterable[str],
+        new_bookmarks: t.Iterable[str]
+    ) -> None:
+        new_bms = set(new_bookmarks)
+        prev_bms = set(previous_bookmarks)
+        with self._lock:
```

Comment:
so you went with a simple lock, right? not a rwlock, it seems, or?

Reply:
Yes, a simple plain lock<sup>*)<sup>.

<sup>*)</sup> The lock is actually a noop lock in the async implementation, because there is not `await` while the lock is acquired: this means there is no possibility of concurrency given that Python's async implements cooperative concurrency. I.e., you never get interrupted, unless you yield to the event loop (some `await` calls).

Furthermore, there is no builtin rwlock in Python. So I would've had to roll my own solution. That definitely seems overkill for the BMM.